app/rdf/query-hierarchies.ts

```diff
@@ -0,0 +1,88 @@
+import { DESCRIBE } from "@tpluscode/sparql-builder";
+import clownface from "clownface";
+import { ascending } from "d3";
+import { uniqBy } from "lodash";
+import rdf from "rdf-ext";
+import { DatasetCore, NamedNode, Quad, Stream } from "rdf-js";
+
+import { HierarchyValue } from "@/graphql/resolver-types";
+
+import * as ns from "./namespace";
+import { sparqlClientStream } from "./sparql-client";
+
+const fromStream = (
+  dataset: DatasetCore<Quad, Quad>,
+  stream: Stream<Quad>
+): Promise<DatasetCore<Quad, Quad>> => {
+  return new Promise((resolve) => {
+    stream.on("data", (quad: Quad) => dataset.add(quad));
+    stream.on("end", () => resolve(dataset));
+  });
+};
+
+const hasValueAndLabel = (
+  o: Record<string, any>
+): o is Record<string, string> & { label: string; value: string } => {
+  return o.label && o.value ? true : false;
+};
+
+export const queryHierarchy = async (
+  dimension: string
+): Promise<HierarchyValue[]> => {
+  const query = DESCRIBE`?hierarchy ?level1 ?level2`.WHERE`    
+      SELECT DISTINCT ?hierarchy ?level1 ?level2
```

Comment:
Why is there a DESCRIBE and then SELECT inside a query? Shouldn't this just be SELECT.DISTINCT`?hierarchy ?level1 ?level2...`? What is the difference when using this notation?

Reply:
I used the same technique as did Tomas from Zazuko here.

First we select the IRIs we are interested in (inside the inner select), and then we describe them to get all the attributes. We end up with an in-memory graph full of interconnected objects, and we use clownface to navigate it and reconstruct the hierarchy.

I *think* we could have a way to get tabular data right away to end up with a table with columns (iri, label, parentIri) and then reconstruct the hierarchy tree, I have to try.

Reply:
I see, thanks for explanation 🙏